sdk/go/pulumi-language-go/main.go

```diff
@@ -339,10 +337,38 @@ func (host *goLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest)
 		return nil, errors.Wrap(err, "failed to prepare environment")
 	}
 
-	cmd, err := findProgram(host.binary)
+	// the user can explicitly opt in to using a binary executable by specifying
+	// runtime.options.binary in the Pulumi.yaml
+	program := host.binary
+	if program == "" {
+		// user did not specify a binary and we will compile and run the binary on-demand
+		logging.V(5).Infof("No prebuilt executable specified, attempting invocation via compilation")
+
+		pulumiHome, err := workspace.GetPulumiHomeDir()
+		if err != nil {
+			return nil, err
+		}
+
+		file, err := ioutil.TempFile(pulumiHome, ".pulumi.out")
+		if err != nil {
+			return nil, errors.Wrap(err, "error in compiling Go")
+		}
+		defer os.Remove(file.Name())
```

Comment:
So the os.Remove here negates the perf benefits asked for in the ticket https://github.com/pulumi/pulumi/issues/10350

Several Pulumi operations on the same program without editing source code will *still* relink the program and be slow in presence of os.Remove.

Why are we including it?


Why is it included?

Reply:
I think there are unexpected risks around how users may be depending on `go run` over `go build` it'd be easier to ship a small change to test the waters and then build on top of it.

I've been told that changes to Go can have unexpected surprises given how our power users use it.